ui/components-react/src/components-react/filter-builder/FilterBuilderState.ts

```diff
@@ -167,34 +170,147 @@ export class PropertyFilterBuilderActions {
       rule.value = value;
     });
   }
+
+  /**
+   * Sets error messages of the rules specified by id.
+   * If rule id is not present in the map, then its error message will be cleared.
+   */
+  public setRuleErrorMessages(ruleIdsAndErrorMessages: Map<string, string>) {
+    this.updateState((state) => {
+      const setErrorMessages = (item: PropertyFilterBuilderRuleGroupItem) => {
+        if (isPropertyFilterBuilderRuleGroup(item)) {
+          item.items.forEach((itm) => {
+            setErrorMessages(itm);
+          });
+        } else {
+          item.errorMessage = ruleIdsAndErrorMessages.get(item.id);
+        }
+      };
+
+      setErrorMessages(state.rootGroup);
+    });
+  }
 }
 
 /**
  * Function to check if supplied [[PropertyFilterBuilderRuleGroupItem]] is [[PropertyFilterBuilderRuleGroup]].
- * @internal
+ * @beta
  */
 export function isPropertyFilterBuilderRuleGroup(
   item: PropertyFilterBuilderRuleGroupItem
 ): item is PropertyFilterBuilderRuleGroup {
   return (item as any).items !== undefined;
 }
 
+/**
+ * Props for [[usePropertyFilterBuilder]]
+ * @beta
+ */
+export interface UsePropertyFilterBuilderProps {
+  initialFilter?: PropertyFilter;
+}
 /**
  * Custom hook that creates state for [[PropertyFilterBuilder]] component. It creates empty state or initializes
  * state from supplied initial filter.
- * @internal
+ * @returns root group of the [[PropertyFilterBuilder]], actions for manipulating state and validate function, which is used for validating rules.
+ * Validate function accepts [[PropertyFilterBuilderRuleGroupItem]] and should return error message or undefined.
+ * @beta
  */
-export function usePropertyFilterBuilderState(initialFilter?: PropertyFilter) {
-  const [state, setState] = React.useState<PropertyFilterBuilderState>(() =>
+export function usePropertyFilterBuilder({
+  initialFilter,
+}: UsePropertyFilterBuilderProps) {
+  const [state, setState] = React.useState<PropertyFilterBuilderState>(
     initialFilter
       ? convertFilterToState(initialFilter)
       : { rootGroup: createEmptyRuleGroup() }
   );
+
   const [actions] = React.useState(
     () => new PropertyFilterBuilderActions(setState)
   );
 
-  return { state, actions };
+  const validate = (
+    ruleValidator?: (item: PropertyFilterBuilderRule) => string | undefined
+  ) => {
+    const ruleIdsAndErrorMessages = new Map<string, string>();
+    ruleGroupItemValidator({
+      actions,
+      item: state.rootGroup,
+      path: [],
+      ruleIdsAndErrorMessages,
+      ruleValidator,
+    });
```

Comment:
I would suggest refactoring this a little bit:
```suggestion
    const ruleErrors = validateRules(state.rootGroup, ruleValidator);
```
`validateRules` can create `ruleErrors` map so you don't need to pass it through parameters.

Reply:
fixed